// @woodwing/studio-component-set-tools/dist/models/component-property-controls.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.COMPONENT_PROPERTY_CONTROL_FITTING_VALUES = exports.isSlider = exports.isUrl = exports.isTextarea = exports.isHeader = exports.isInteractive = exports.isSlides = exports.isFitting = exports.isMediaProperties = exports.isDropCapital = exports.isImageEditor = exports.isColorPicker = exports.isTime = exports.isText = exports.isDisableFullscreenCheckbox = exports.isCheckbox = exports.isRadio = exports.isSelect = void 0;
function isSelect(control) {
    return control.type === 'select';
}
exports.isSelect = isSelect;
function isRadio(control) {
    return control.type === 'radio';
}
exports.isRadio = isRadio;
function isCheckbox(control) {
    return control.type === 'checkbox';
}
exports.isCheckbox = isCheckbox;
function isDisableFullscreenCheckbox(control) {
    return control.type === 'disable-fullscreen-checkbox';
}
exports.isDisableFullscreenCheckbox = isDisableFullscreenCheckbox;
function isText(control) {
    return control.type === 'text';
}
exports.isText = isText;
function isTime(control) {
    return control.type === 'time';
}
exports.isTime = isTime;
function isColorPicker(control) {
    return control.type === 'colorPicker';
}
exports.isColorPicker = isColorPicker;
function isImageEditor(control) {
    return control.type === 'image-editor';
}
exports.isImageEditor = isImageEditor;
function isDropCapital(control) {
    return control.type === 'drop-capital';
}
exports.isDropCapital = isDropCapital;
function isMediaProperties(control) {
    return control.type === 'media-properties';
}
exports.isMediaProperties = isMediaProperties;
function isFitting(control) {
    return control.type === 'fitting';
}
exports.isFitting = isFitting;
function isSlides(control) {
    return control.type === 'slides';
}
exports.isSlides = isSlides;
function isInteractive(control) {
    return control.type === 'interactive';
}
exports.isInteractive = isInteractive;
function isHeader(control) {
    return control.type === 'header';
}
exports.isHeader = isHeader;
function isTextarea(control) {
    return control.type === 'textarea';
}
exports.isTextarea = isTextarea;
function isUrl(control) {
    return control.type === 'url';
}
exports.isUrl = isUrl;
function isSlider(control) {
    return control.type === 'slider';
}
exports.isSlider = isSlider;
/**
 * Fitting property possible values
 */
exports.COMPONENT_PROPERTY_CONTROL_FITTING_VALUES = {
    FRAME_HEIGHT_TO_CONTENT_TYPE: '_fit-frame-height-to-content',
    FRAME_TO_CONTENT_TYPE: '_fit-frame-to-content',
};
